src/binary/ibc-setup/commands/balances.ts

```diff
@@ -0,0 +1,74 @@
+import path from 'path';
+
+import { GasPrice } from '@cosmjs/launchpad';
+import { Decimal } from '@cosmjs/math';
+import { DirectSecp256k1HdWallet } from '@cosmjs/proto-signing';
+
+import { IbcClient } from '../../../lib/ibcclient';
+import { registryFile } from '../../constants';
+import { getDefaultHomePath } from '../utils/get-default-home-path';
+import { loadAndValidateApp } from '../utils/load-and-validate-app';
+import { loadAndValidateRegistry } from '../utils/load-and-validate-registry';
+import { resolveMnemonicOption } from '../utils/resolve-mnemonic-option';
+import { resolveOption } from '../utils/resolve-option';
+import { resolveRequiredOption } from '../utils/resolve-required-option';
+
+import { Flags, getAddresses, Options } from './keys-list';
+
+export async function balances(flags: Flags) {
+  const home = resolveRequiredOption('home')(
+    flags.home,
+    process.env.RELAYER_HOME,
+    getDefaultHomePath
+  );
+  const app = loadAndValidateApp(home);
+
+  const keyFile = resolveOption(
+    flags.keyFile,
+    app?.keyFile,
+    process.env.KEY_FILE
+  );
+
+  const options: Options = {
+    home,
+    mnemonic: await resolveMnemonicOption({
+      interactive: flags.interactive,
+      mnemonic: flags.mnemonic,
+      keyFile: keyFile,
+      app,
+    }),
+  };
+
+  await run(options);
+}
+
+export async function run(options: Options) {
+  const registryFilePath = path.join(options.home, registryFile);
+  const registry = loadAndValidateRegistry(registryFilePath);
+
+  const addresses = await getAddresses(registry.chains, options.mnemonic);
+
+  for (const [chain, data, address] of addresses) {
+    const signer = await DirectSecp256k1HdWallet.fromMnemonic(
+      options.mnemonic,
+      undefined,
+      data.prefix
+    );
+
+    const gasPrice = GasPrice.fromString(data.gas_price);
+
+    const client = await IbcClient.connectWithSigner(
+      data.rpc[0], // rpc[0] is guaranteed to be defined by registry validator
+      signer,
+      address,
+      {
+        prefix: data.prefix,
+        gasPrice,
+      }
+    );
+
+    const balances = await client.query.bank.unverified.allBalances(address);
```

Comment:
It returns `Coin[]` type.

Should it be iterated and summed?

---

The spec for the `balances` command says:
>prints out all chains for which the relayer has a non-zero balance

does it mean that we need at least one non-zero entry in the `Coin[]` array to include the address in the output?

Reply:
Will can you provide a coinToString function for output?

Reply:
Maybe we only query the feeDenom, which is the only one that is there

Reply:
there is a `bank.unverified.balance(address, denom)` function

Reply:
Looks like we don't have one in CosmJS, but you can see the usual format here: https://github.com/cosmos/cosmjs/blob/72b7c0b/packages/launchpad/src/coins.ts#L19

Reply:
I think something went wrong here. I have the parsing already with `GasPrice`: https://github.com/confio/ts-relayer/pull/91/files#diff-58673c679ff81312a0f7389a4cf697deac398637a352094bd9d4832a5cf28ee4R53-R67

Does it look okay?